### ISSUE TITLE
Segmentation result interpretation

### ISSUE THREAD

**Issue by ffamoss** [ultralytics/ultralytics#1244]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I'm train custom segmentation model with 1 label, check with standard predict function, all works fine.
Then export to onnx format, and using model in С#.
As result, i got two tensors 
1) [1,37,8400] - this is bound boxes and predictions, all fine
2) [1,32,160,160] - this is downscale masks, but i can't find any description about it structure.

I find source code to convert all 32 masks with prediction from output 1 to binary mask.
Then using it to detect polygin by  openCV.

But i can't understand how it works and what is each of 32 mask's meen.
Can you explain step by step what can i do to build a single binary mask? (or any link with algorithm description)

    process_mask(protos, masks_in, bboxes, shape, upsample=False)
        c, mh, mw = protos.shape  # CHW
        ih, iw = shape
        masks = (masks_in @ protos.float().view(c, -1)).sigmoid().view(-1, mh, mw)  # CHW
    
        downsampled_bboxes = bboxes.clone()
        downsampled_bboxes[:, 0] *= mw / iw
        downsampled_bboxes[:, 2] *= mw / iw
        downsampled_bboxes[:, 3] *= mh / ih
        downsampled_bboxes[:, 1] *= mh / ih
    
        masks = crop_mask(masks, downsampled_bboxes)  # CHW
        if upsample:
            masks = F.interpolate(masks[None], shape, mode='bilinear', align_corners=False)[0]  # CHW
        return masks.gt_(0.5)

**Comment by Laughing-q** [comment]
@ffamoss the design of the segmentation model inspired by yolact. you might want to take a look at the paper. :) [https://arxiv.org/pdf/1904.02689v2.pdf](https://arxiv.org/pdf/1904.02689v2.pdf)

**Comment by ffamoss** [comment]
> @ffamoss the design of the segmentation model inspired by yolact. you might want to take a look at the paper. :) https://arxiv.org/pdf/1904.02689v2.pdf

Thank you.
The algorithm is clear to me.
May be helpfull

I add copy mask prediction from output0 see:
https://github.com/sstainba/Yolov8.Net/blob/main/src/Yolov8net/Yolov8.cs
```
 var maskPred = Enumerable.Range(5, output.Dimensions[1] - 5)
                                    .Select(m => output[i, m, j])
                                    .ToArray();
```

Code that visualizes each step to make debugging easier. (Please forgive my coding style; it was an experiment:) )

```
        public float[,] MultiplyMatrix(Tensor<float> A, float[][] B, Func<float,float> elemFunc)
        {
            int rA = A.Dimensions[2];
            int cA = A.Dimensions[1];
            int rB = B[0].Length;
            int cB = B.Length;

            if (cA != rB)
            {
                throw new ArgumentOutOfRangeException("Matrixes can't be multiplied!!");
            }
            else
            {
                float temp = 0;
                float[,] kHasil = new float[rA, cB];

                for (int i = 0; i < rA; i++)
                {
                    for (int j = 0; j < cB; j++)
                    {
                        temp = 0;
                        for (int k = 0; k < cA; k++)
                        {
                            temp += A[0, k, i] * B[j][k];
                        }
                        kHasil[i, j] = elemFunc(temp);
                    }
                }

                return kHasil;
            }
        }

        private void PrintOut1(List<YoloPrediction> boxes, DenseTensor<float> det, DenseTensor<float> output, Image image)
        {
            var result = new ConcurrentBag<YoloPrediction>();
            var shape = 160;
            var dir = @"\output";
            var (w, h) = (image.Width, image.Height); // image w and h
            var (xGain, yGain) =( shape / (float)w, shape / (float)h); // x, y gains
            var gain = Math.Min(xGain, yGain); // gain = resized / original
                       

            var r = output.Reshape(new[] { 1, 32, 160 * 160 });
            var mask = MultiplyMatrix(r,boxes.Select(m=>m.MaskPrediction).ToArray(), x=> Sigmoid(x) >= 0.5? 1:0);

            var resultImg = ((Bitmap)image).ToImage<Bgr, byte>();

            for (int i = 0; i < mask.GetLength(1); i++)
            {
                var btm = new Bitmap(shape, shape);
                var box = boxes[i].Rectangle;
                box = new RectangleF(box.X * xGain, box.Y * yGain, box.Width * xGain, box.Height * yGain);


                for (int j = 0; j < mask.GetLength(0); j++)
                {
                    var color = mask[j, i] == 0 ? Color.White : Color.Black;
                    int y = j / shape; // Integer division 
                    int x = j % shape;
                    if (box.Contains(x, y))
                    {
                        btm.SetPixel(x, y, color);
                    }
                    else
                    {
                        btm.SetPixel(x, y, Color.White);
                    }
                }

                
                btm.Save($"{dir}\\{i}.jpg");

                var fullMask = btm.ToImage<Gray, byte>().Resize(640, 640, Emgu.CV.CvEnum.Inter.LinearExact);
                fullMask.Save($"{dir}\\{i}_f.jpg");

                VectorOfVectorOfPoint contours = new VectorOfVectorOfPoint();
                Mat hier = new Mat();
                
                CvInvoke.FindContours(fullMask, contours, hier, Emgu.CV.CvEnum.RetrType.Ccomp, Emgu.CV.CvEnum.ChainApproxMethod.ChainApproxSimple);
                CvInvoke.DrawContours(resultImg, contours, -1, new MCvScalar(255, 0, 0), 2);

                var fullMaskC = fullMask.Convert<Bgr, byte>();
                CvInvoke.DrawContours(fullMaskC, contours, -1, new MCvScalar(255, 0, 0), 2);
                fullMaskC.Save($"{dir}\\{i}_c.jpg");
            }


            resultImg.Save($"{dir}\\result.jpg");
        }
```
